### PR TITLE
[FLINK-16743][table] Introduce datagen, print, blackhole connectors

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Introduce built-in connectors for better startup, test, production-debug, and etc...

1.datagen source:
- easy startup/test for streaming job
- performance testing

2.print sink:
- easy test for streaming job
- be very useful in production debugging

3.blackhole sink
- very useful for high performance testing of Flink
- I've also run into users trying UDF to output, not sink, so they need this sink as well.

## Brief change log

Introduce:
- DataGeneratorSource
- DataGenTableSourceFactory
- PrintTableSinkFactory
- BlackHoleTableSinkFactory

## Verifying this change

- `DataGenTableSourceFactoryTest`
- `TableSinkFactoryTest`
- `DataGeneratorSourceTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by shangwen** [comment]
hi @JingsongLi , Could you please review this PR for me?

**Comment by bowenli86** [comment]
revive the thread. These will be very useful for users

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	 * @param rowsPerSecond Control the emit rate.
+	 */
+	public DataGeneratorSource(DataGenerator<T> generator, long rowsPerSecond) {
+		this.generator = generator;
+		this.rowsPerSecond = rowsPerSecond;
+	}
+
+	@Override
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+		this.generator.open("TOP_FIELD", context, getRuntimeContext());
```
Comment (bowenli86) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java:64]: where does the "TOP_FIELD" come from?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		double taskRowsPerSecond = (double) rowsPerSecond / getRuntimeContext().getNumberOfParallelSubtasks();
+		long readTimeIncrement = (long) (1000 / taskRowsPerSecond);
+		long nextReadTime = System.currentTimeMillis();
+
+		while (isRunning && generator.hasNext()) {
+			synchronized (ctx.getCheckpointLock()) {
+				ctx.collect(this.generator.next());
+			}
+
+			if (readTimeIncrement > 0) {
```
Comment (bowenli86) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java:84]: what if I want 10,000 and 20,000 rps with single parallelism? How does the rate control work?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/RandomGenerator.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+			FunctionInitializationContext context,
+			RuntimeContext runtimeContext) throws Exception {
+		this.random = new RandomDataGenerator();
+	}
+
+	@Override
+	public boolean hasNext() {
+		return true;
+	}
+
```
Comment (bowenli86) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/RandomGenerator.java:47]: a few small feature requests: add decimal, and an array generator with the existing primitive types?

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/BlackHoleTableSinkFactory.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_PROPERTY_VERSION;
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_TYPE;
+import static org.apache.flink.table.descriptors.DescriptorProperties.TABLE_SCHEMA_EXPR;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA;
+
+/**
+ * Factory base for creating configured instances of {@link CsvTableSink} in a stream environment.
```
Comment (bowenli86) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/BlackHoleTableSinkFactory.java:42]: update comment?

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/PrintTableSinkFactory.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_PROPERTY_VERSION;
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_TYPE;
+import static org.apache.flink.table.descriptors.DescriptorProperties.TABLE_SCHEMA_EXPR;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA;
+
+/**
+ * Factory base for creating configured instances of {@link CsvTableSink} in a stream environment.
```
Comment (bowenli86) [inline flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/sinks/PrintTableSinkFactory.java:42]: update comment?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java (hunk 1 above)
Comment (JingsongLi) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java:64]: This is just name for generator, I will change to "DataGenerator".

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java (hunk 2 above)
Comment (JingsongLi) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/DataGeneratorSource.java:84]: I'll change to nanos.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/RandomGenerator.java (hunk 3 above)
Comment (JingsongLi) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/RandomGenerator.java:47]: We don't have a good generator for decimal and array.

**Comment by JingsongLi** [comment]
Hi @bowenli86 , DataGen part was moved into https://github.com/apache/flink/pull/12074 , FYI

**Pull request closed without merging**